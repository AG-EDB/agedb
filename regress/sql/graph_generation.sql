/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */


LOAD 'age';

SET search_path = ag_catalog;

SELECT * FROM create_complete_graph('gp1',5,'edges','vertices');

SELECT COUNT(*) FROM gp1."edges";
SELECT COUNT(*) FROM gp1."vertices";

SELECT * FROM cypher('gp1', $$MATCH (a)-[e]->(b) RETURN e$$) as (n agtype);

SELECT * FROM create_complete_graph('gp1',5,'edges','vertices');

SELECT COUNT(*) FROM gp1."edges";
SELECT COUNT(*) FROM gp1."vertices";

SELECT * FROM create_complete_graph('gp2',5,'edges');

SELECT * FROM create_complete_graph('gp3',5, NULL);

SELECT * FROM create_complete_graph('gp4',NULL,NULL);

SELECT * FROM create_complete_graph(NULL,NULL,NULL);

SELECT drop_graph('gp1', true);
SELECT drop_graph('gp2', true);


-- PATH GRAPH TESTS

-- Should create 5 vertices with 4 edges. All standard ag_labels.
SELECT age_create_path('path_graph_5_false', 5, 'PathVertex', 'PATH_EDGE', false);
SELECT * FROM path_graph_5_false."PathVertex";
SELECT * FROM path_graph_5_false."PATH_EDGE";

-- Should create 5 vertices with 8 edges. All standard ag_labels.
SELECT age_create_path('path_graph_5_true', 5, 'PathVertex', 'PATH_EDGE', true);
SELECT * FROM path_graph_5_true."PathVertex";
SELECT * FROM path_graph_5_true."PATH_EDGE";

-- Should create 10 vertices with 9 edges. New ag_labels.
SELECT age_create_path('path_graph_10_false', 10, 'PathVertex', 'PATH_EDGE', false);
SELECT * FROM path_graph_10_false."PathVertex";
SELECT * FROM path_graph_10_false."PATH_EDGE";

-- Should create 10 vertices with 18 edges. New ag_labels.
SELECT age_create_path('path_graph_10_true', 10, 'PathVertex', 'PATH_EDGE', true);
SELECT * FROM path_graph_10_true."PathVertex";
SELECT * FROM path_graph_10_true."PATH_EDGE";

-- Should throw errors.
SELECT age_create_path('path_graph_1', 1, 'PathVertex', 'PATH_EDGE', false);
SELECT age_create_path(NULL, NULL, NULL, NULL, NULL);

-- Drop Path Graphs.
SELECT drop_graph('path_graph_5_false', true);
SELECT drop_graph('path_graph_5_true', true);
SELECT drop_graph('path_graph_10_false', true);
SELECT drop_graph('path_graph_10_true', true);