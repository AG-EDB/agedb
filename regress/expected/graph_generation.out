/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
LOAD 'age';
SET search_path = ag_catalog;
SELECT * FROM create_complete_graph('gp1',5,'edges','vertices');
NOTICE:  graph "gp1" has been created
NOTICE:  VLabel "vertices" has been created
NOTICE:  ELabel "edges" has been created
 create_complete_graph 
-----------------------
 
(1 row)

SELECT COUNT(*) FROM gp1."edges";
 count 
-------
    10
(1 row)

SELECT COUNT(*) FROM gp1."vertices";
 count 
-------
     5
(1 row)

SELECT * FROM cypher('gp1', $$MATCH (a)-[e]->(b) RETURN e$$) as (n agtype);
                                                             n                                                              
----------------------------------------------------------------------------------------------------------------------------
 {"id": 1125899906842625, "label": "edges", "end_id": 844424930131970, "start_id": 844424930131969, "properties": {}}::edge
 {"id": 1125899906842629, "label": "edges", "end_id": 844424930131971, "start_id": 844424930131970, "properties": {}}::edge
 {"id": 1125899906842626, "label": "edges", "end_id": 844424930131971, "start_id": 844424930131969, "properties": {}}::edge
 {"id": 1125899906842630, "label": "edges", "end_id": 844424930131972, "start_id": 844424930131970, "properties": {}}::edge
 {"id": 1125899906842627, "label": "edges", "end_id": 844424930131972, "start_id": 844424930131969, "properties": {}}::edge
 {"id": 1125899906842632, "label": "edges", "end_id": 844424930131972, "start_id": 844424930131971, "properties": {}}::edge
 {"id": 1125899906842631, "label": "edges", "end_id": 844424930131973, "start_id": 844424930131970, "properties": {}}::edge
 {"id": 1125899906842634, "label": "edges", "end_id": 844424930131973, "start_id": 844424930131972, "properties": {}}::edge
 {"id": 1125899906842633, "label": "edges", "end_id": 844424930131973, "start_id": 844424930131971, "properties": {}}::edge
 {"id": 1125899906842628, "label": "edges", "end_id": 844424930131973, "start_id": 844424930131969, "properties": {}}::edge
(10 rows)

SELECT * FROM create_complete_graph('gp1',5,'edges','vertices');
 create_complete_graph 
-----------------------
 
(1 row)

SELECT COUNT(*) FROM gp1."edges";
 count 
-------
    20
(1 row)

SELECT COUNT(*) FROM gp1."vertices";
 count 
-------
    10
(1 row)

SELECT * FROM create_complete_graph('gp2',5,'edges');
NOTICE:  graph "gp2" has been created
NOTICE:  ELabel "edges" has been created
 create_complete_graph 
-----------------------
 
(1 row)

SELECT * FROM create_complete_graph('gp3',5, NULL);
ERROR:  edge label can not be NULL
SELECT * FROM create_complete_graph('gp4',NULL,NULL);
ERROR:  number of nodes can not be NULL
SELECT * FROM create_complete_graph(NULL,NULL,NULL);
ERROR:  graph name can not be NULL
SELECT drop_graph('gp1', true);
NOTICE:  drop cascades to 4 other objects
DETAIL:  drop cascades to table gp1._ag_label_vertex
drop cascades to table gp1._ag_label_edge
drop cascades to table gp1.vertices
drop cascades to table gp1.edges
NOTICE:  graph "gp1" has been dropped
 drop_graph 
------------
 
(1 row)

SELECT drop_graph('gp2', true);
NOTICE:  drop cascades to 3 other objects
DETAIL:  drop cascades to table gp2._ag_label_vertex
drop cascades to table gp2._ag_label_edge
drop cascades to table gp2.edges
NOTICE:  graph "gp2" has been dropped
 drop_graph 
------------
 
(1 row)

-- Tests for the Tadpole graph generation algorithm.
SELECT * FROM ag_catalog.age_create_tadpole_graph('TadpoleGraphOne', 3, 4, 'vertex', 'edge', false);
NOTICE:  graph "TadpoleGraphOne" has been created
NOTICE:  VLabel "vertex" has been created
NOTICE:  ELabel "edge" has been created
 age_create_tadpole_graph 
--------------------------
 
(1 row)

SELECT * FROM ag_catalog.age_create_tadpole_graph('TadpoleGraphTwo', 4, 2, 'vertex', 'edge', true);
NOTICE:  graph "TadpoleGraphTwo" has been created
NOTICE:  VLabel "vertex" has been created
NOTICE:  ELabel "edge" has been created
 age_create_tadpole_graph 
--------------------------
 
(1 row)

SELECT * FROM cypher('TadpoleGraphOne', $$ MATCH (a)-[e]->(b) RETURN e $$) as (n agtype);
                                                             n                                                             
---------------------------------------------------------------------------------------------------------------------------
 {"id": 1125899906842627, "label": "edge", "end_id": 844424930131969, "start_id": 844424930131971, "properties": {}}::edge
 {"id": 1125899906842625, "label": "edge", "end_id": 844424930131970, "start_id": 844424930131969, "properties": {}}::edge
 {"id": 1125899906842626, "label": "edge", "end_id": 844424930131971, "start_id": 844424930131970, "properties": {}}::edge
 {"id": 1125899906842628, "label": "edge", "end_id": 844424930131972, "start_id": 844424930131971, "properties": {}}::edge
 {"id": 1125899906842629, "label": "edge", "end_id": 844424930131973, "start_id": 844424930131972, "properties": {}}::edge
 {"id": 1125899906842630, "label": "edge", "end_id": 844424930131974, "start_id": 844424930131973, "properties": {}}::edge
 {"id": 1125899906842631, "label": "edge", "end_id": 844424930131975, "start_id": 844424930131974, "properties": {}}::edge
(7 rows)

SELECT * FROM cypher('TadpoleGraphTwo', $$ MATCH (a)-[e]->(b) RETURN e $$) as (n agtype);
                                                             n                                                             
---------------------------------------------------------------------------------------------------------------------------
 {"id": 1125899906842626, "label": "edge", "end_id": 844424930131969, "start_id": 844424930131970, "properties": {}}::edge
 {"id": 1125899906842631, "label": "edge", "end_id": 844424930131969, "start_id": 844424930131972, "properties": {}}::edge
 {"id": 1125899906842625, "label": "edge", "end_id": 844424930131970, "start_id": 844424930131969, "properties": {}}::edge
 {"id": 1125899906842628, "label": "edge", "end_id": 844424930131970, "start_id": 844424930131971, "properties": {}}::edge
 {"id": 1125899906842627, "label": "edge", "end_id": 844424930131971, "start_id": 844424930131970, "properties": {}}::edge
 {"id": 1125899906842630, "label": "edge", "end_id": 844424930131971, "start_id": 844424930131972, "properties": {}}::edge
 {"id": 1125899906842634, "label": "edge", "end_id": 844424930131972, "start_id": 844424930131973, "properties": {}}::edge
 {"id": 1125899906842632, "label": "edge", "end_id": 844424930131972, "start_id": 844424930131969, "properties": {}}::edge
 {"id": 1125899906842629, "label": "edge", "end_id": 844424930131972, "start_id": 844424930131971, "properties": {}}::edge
 {"id": 1125899906842633, "label": "edge", "end_id": 844424930131973, "start_id": 844424930131972, "properties": {}}::edge
 {"id": 1125899906842636, "label": "edge", "end_id": 844424930131973, "start_id": 844424930131974, "properties": {}}::edge
 {"id": 1125899906842635, "label": "edge", "end_id": 844424930131974, "start_id": 844424930131973, "properties": {}}::edge
(12 rows)

SELECT COUNT(*) FROM "TadpoleGraphOne"."edge"; -- Should return 7.
 count 
-------
     7
(1 row)

SELECT COUNT(*) FROM "TadpoleGraphTwo"."edge"; -- Should return 12.
 count 
-------
    12
(1 row)

SELECT * FROM ag_catalog.age_create_tadpole_graph('TadpoleGraphTwo', 5, 1, 'vertex', 'edge', true); -- Creates another one in the same namespace.
 age_create_tadpole_graph 
--------------------------
 
(1 row)

SELECT COUNT(*) FROM "TadpoleGraphTwo"."edge"; -- Should return 24.
 count 
-------
    24
(1 row)

-- Should throw errors.
SELECT * FROM ag_catalog.age_create_tadpole_graph(NULL, 4, 2, 'vertex', 'edge', true);
ERROR:  Graph name cannot be NULL
SELECT * FROM ag_catalog.age_create_tadpole_graph('TadpoleError', 2, 2, 'vertex', 'edge', true);
NOTICE:  graph "TadpoleError" has been created
ERROR:  Number of vertices for the cycle portion cannot be NULL or less than 3.
SELECT * FROM ag_catalog.age_create_tadpole_graph('TadpoleError', 3, 0, 'vertex', 'edge', true);
NOTICE:  graph "TadpoleError" has been created
ERROR:  Number of vertices for the path portion cannot be NULL or less than 1.
-- Drop graphs
SELECT drop_graph('TadpoleGraphOne', true);
NOTICE:  drop cascades to 4 other objects
DETAIL:  drop cascades to table "TadpoleGraphOne"._ag_label_vertex
drop cascades to table "TadpoleGraphOne"._ag_label_edge
drop cascades to table "TadpoleGraphOne".vertex
drop cascades to table "TadpoleGraphOne".edge
NOTICE:  graph "TadpoleGraphOne" has been dropped
 drop_graph 
------------
 
(1 row)

SELECT drop_graph('TadpoleGraphTwo', true);
NOTICE:  drop cascades to 4 other objects
DETAIL:  drop cascades to table "TadpoleGraphTwo"._ag_label_vertex
drop cascades to table "TadpoleGraphTwo"._ag_label_edge
drop cascades to table "TadpoleGraphTwo".vertex
drop cascades to table "TadpoleGraphTwo".edge
NOTICE:  graph "TadpoleGraphTwo" has been dropped
 drop_graph 
------------
 
(1 row)

